### PR TITLE
#294 Better Table Loading Approach.

### LINKED ISSUE #294: Better Table Loading Approach
As per [this](https://github.com/self-xdsd/self-web/pull/293#discussion_r549733804) comment, we should have a more elegant approach to loading data into tables.

```javascript
$("#table").dataTable({
     language: { loadingRecords: '<img src=...>' },
     ajax: function(_, callback){
         callRemote(...).then(remoteData => {
             callback({data: remoteData.map(toTableRow)})
         })
     }
});
```

Let's start by modifying Tasks and Invoices tables on the Project page, then make the same changes for the tables on the Contributor Dashboard page.

### PR DESCRIPTION
**Pull request by criske**
Closes #294
Self XDSD

### CODE REVIEW

**Comment by charlesmike** [comment]
@criske thank you for your Pull Request. I'll assign someone to review it soon.

**Comment by charlesmike** [comment]
@amihaiemil please review this Pull Request. Deadline (when it should be merged or closed) is ``2021-01-02T13:57:06.333824``.

You should check if the requirements have been implemented (partially or in full), if there are unit tests covering the changes and if the CI build passes. Feel free to reject the PR or ask for changes if it's too big or not clear enough.

Estimation here is ``30 minutes``, that's how much you will be paid. You will be paid even if this PR gets rejected.

File: src/main/resources/public/js/getAndAddContracts.js [hunk 1]
```diff
@@ -20,165 +20,147 @@ var projectContractsCount = -1;
...
+                                + data;
+                        } else if (contract.id.provider == 'gitlab') {
+                            issueLink = 'https://gitlab.com/'
+                                + contract.id.repoFullName
+                                + "/issues/"
+                                + data;
+                        } else {
+                            issueLink = '#';
+                        }
+                        return "<a href='" + issueLink + "' target='_blank'>#" + data + "</a>"
```
Comment (criske) [inline src/main/resources/public/js/getAndAddContracts.js:56]: @amihaiemil I think most of this formatting can be done on server, on json creation.
`task` json might look like
```javascript
{
  idTask: { id: '#1', link: 'http://github...'},
  assignmentDate: <formatted date>,
  deadline: <formatted date>,
  estimation: <formattted estimation>,
  value: <euro formatted value>
}
```
This way we only need to render in table the `a` element for the link.

File: src/main/resources/public/js/getAndAddContracts.js [hunk 2]
```diff
@@ -20,165 +20,147 @@ var projectContractsCount = -1;
...
-        var downloadLink = "<a href='"+pdfHref+"' title='Download Invoice' class='downloadInvoice'>"
-            + "<i class='fa fa-file-pdf-o fa-lg'></i>"
-            + "</a>  ";
-        if (invoice.paymentTime == "null" && invoice.transactionId == "null") {
-            status = "Active";
-            if (parseFloat(invoice.totalAmount) > 0.0) {
-                payIcon = "<a href='#' title='Pay Invoice' class='payInvoice'>"
-                    + "<i class='fa fa-credit-card fa-lg'></i>"
-                    + "</a>";
+    function invoiceAsTableRow(contract) {
```
Comment (criske) [inline src/main/resources/public/js/getAndAddContracts.js:134]: @amihaiemil Same, the presentation can be done on the server.
```javascript
{
 id: 1,
 createdAt: "formatted date",
 amount: "formatted euro",
 status: "Paid|Active",
 options: {
   downloadLink : "<endpoint link>",
   paymentLink: "<endpoint link>"
 }
}
```
For options when rendering, we  only need to check if download/payment link is empty. If they are not empty then we attach the click listeners and showing icons, as usual.
Download/Payment link will be empty if amount is less than 108 euro, so we will avoid checking this on frontend.

**Comment by amihaiemil** [comment]
@criske yes, the ideas are good, but let's leave it for later. We will prepare the first Prod release soon, so I want to limit the new changes at the moment :D 

Also, can you write a puzzle to continue making these table changes on the Contributor Dashboard page as well?

**Comment by charlesmike** [comment]
@amihaiemil thank you for resolving this ticket. I've just added it to your active invoice. You can always check all your invoices and more on the [Contributor Dashboard](https://self-xdsd.com).

**Pull request merged**